crates/matrix-sdk/src/room/common.rs

```diff
@@ -194,6 +198,97 @@ impl Common {
         Ok(response)
     }
 
+    /// Get a stream for the timeline of this `Room`
+    ///
+    /// The first stream is forward in time and second stream is backward in
+    /// time. If the `Store` used implements message caching and the
+    /// timeline is cached no request to the server is made.
+    ///
+    /// The streams make sure that no duplicated events are returned. If the
+    /// event graph changed on the server existing streams will keep the
+    /// previous event order. Streams created later will use the new event
+    /// graph, and therefore will create a new local cache.
+    ///
+    /// The forward stream will only return `None` when a gapped sync was
+    /// performed. In this case also the backward stream should be dropped,
+    /// to make the best use of the message cache.
+    ///
+    /// The backward stream returns `None` once the first event of the room was
+    /// reached. The backward stream may also return an Error when a request
+    /// to the server failed. If the error is persistent new streams need to
+    /// be created.
+    ///
+    /// With the encryption feature, messages are decrypted if possible. If
+    /// decryption fails for an individual message, that message is returned
+    /// undecrypted.
+    ///
+    /// # Examples
+    /// ```no_run
+    /// # use std::convert::TryFrom;
+    /// use matrix_sdk::{room::MessagesOptions, Client};
+    /// # use matrix_sdk::ruma::{
+    /// #     api::client::r0::filter::RoomEventFilter,
+    /// #     room_id,
+    /// # };
+    /// # use url::Url;
+    ///
+    /// # let homeserver = Url::parse("http://example.com").unwrap();
+    /// # use futures::executor::block_on;
+    /// # block_on(async {
+    ///
+    /// let mut client = Client::new(homeserver).await.unwrap();
+    /// let room = client
+    ///    .get_joined_room(room_id!("!roomid:example.com"))
+    ///    .unwrap();
+    /// assert!(room.timeline().await.is_ok());
+    /// # });
+    /// ```
+    pub async fn timeline(
```

Comment:
> Wouldn't you instead want a single stream that you can move forwards and backwards? I guess there's no existing standard trait for such a behavior.

The initial idea was to use https://docs.rs/async-std/latest/async_std/stream/trait.DoubleEndedStream.html

> Another thing that comes to mind, isn't this yet another way to get events from a sync for a room? We have a sync callback API, a sync stream API, the event handler and now a unbounded channel per room we can stream events from? Not that we can't have this one, but I wonder if the streaming sync API is unused, AFAIR you requested it.

Yes it is. But specific to the timeline that gives you all events of a timeline not only sync responses. Btw we are moving away from the streaming sync API, towards using the event handler which make much more sense for us. I didn't ask for the  streaming sync API, but we use it because it was a little nice then the `with_callback` API.